### ISSUE TITLE
[Question] imports insed less works?

### ISSUE THREAD

**Issue by gullitmiranda** [jean343/styless#29]
I'm trying to use to import a file that has import. but does not work. it's supported?

styless:

```js
const HeaderStyled = styled(Layout.Header)`
  @import (reference) "tmp.less";
  &.light {
    color: @text-color;
    background: @layout-sider-background-light;
  }

  &.dark {
    background: @layout-header-background;
    color: @layout-trigger-color;
  }
`;
```

less:

```less
/* tmp.less */
@text-color: red;
@import "test.less";

/* test.less */
@import "test2.less";
@test: red;
.test {
  color: @test;
}

@text-color: cyan;

/* test2.less */
@test: red;
.test2 {
  color: @test;
}
```

**Comment by jean343** [comment]
Thanks for the detailed test case, I will add it as a test soon.
Nested imports are supported,
In this example, the selectors `.test` and `.test2` are not injected because `tmp.less` is imported as a `(reference)`.
Also, the class `.dark` is not generated since the variables `@layout-header-background` and `@layout-trigger-color` are not defined.

Let me know if this helps.

**Comment by jean343** [comment]
I did not mean to close the defect by merging my test addition.
If you need more information, feel free to re-open it.

Please note that you may need to clean your babel cache when editing the source less files.
I added the command `yarn clean` to clear the jest cache.

If the variables are meant to be dynamic, I prefer setting them in the theme store rather than a less import. See [this](https://github.com/jean343/styless/blob/master/docs/receipe-theme.md) for more info.

**Comment by gullitmiranda** [comment]
I will test without `(reference)` today later. 
the variables are not very dynamic, but I'm working on a way to charge them less, supporting the import's with a custom webpack loader:

```js
const less = require("less");
const getOptions = require("less-loader/dist/getOptions");
const lessToJs = require("./less-vars-to-js");

const parseResult = (context, options) => {
  const promise = new Promise(function(resolve, reject) {
    less.parse(context, options, function(err, root, imports) {
      if (err) reject(err);
      else resolve({ root, imports });
    });
  });

  return promise;
};

module.exports = async function(context) {
  const loaderContext = this;
  const options = getOptions(loaderContext);
  let results = [];
  lessToJs;

  const lessToJsOptions = {
    resolveVariables: true,
    stripPrefix: true
  };

  const {
    // root,
    imports
  } = await parseResult(context, options);

  results = Object.keys(imports.contents).map(key => imports.contents[key]);
  results.reverse();

  let lessContent = results.join("\n");

  lessContent = lessContent.replace(/@import.[^;]*;/gm, "");
  const lessJs = lessToJs(lessContent, lessToJsOptions);

  return `module.exports = ${JSON.stringify(lessJs)}`;
};
```

and

```jsx
import React from "react";
import { ThemeProvider as StyledThemeProvider } from "styled-components";

/* eslint-disable */
import * as themeDefault from "!!../utils/less-var-loader!./default/index.less";
/* eslint-enable */

import "./index.less";

const theme = {
  ...themeDefault
};

const ThemeProvider = ({ children }) => {
  return <StyledThemeProvider theme={theme}>{children}</StyledThemeProvider>;
};

export default ThemeProvider;

```

but this don't work properly. I'm thinking about getting the variables from the less tree instead of the parsed string with `lessToJs`.

**Comment by jean343** [comment]
I agree that `less-vars-to-js` is limited as it search for variables in one file only.
The less compiler itself can scan through imports and find all variables.

In your code above, what did you get for `themeDefault`?

**Comment by gullitmiranda** [comment]
a little sample:

```json
{
  "functions": "~`(function() {\n  var hueStep = 2",
  "blue-1": "color(~`colorPalette(\"@{blue-6}\", 1)`)",
  "blue-2": "color(~`colorPalette(\"@{blue-6}\", 2)`)",
  "blue-3": "color(~`colorPalette(\"@{blue-6}\", 3)`)",
  "blue-4": "color(~`colorPalette(\"@{blue-6}\", 4)`)",
  "blue-5": "color(~`colorPalette(\"@{blue-6}\", 5)`)",
  "slider-handle-color": "color(~`colorPalette(\"@{primary-color}\", 3)`)",
  "slider-handle-color-hover": "color(~`colorPalette(\"@{primary-color}\", 4)`)",
  "slider-handle-color-focus": "tint(@primary-color, 20%)",
  "slider-handle-color-focus-shadow": "tint(@primary-color, 50%)",
  "slider-handle-color-tooltip-open": "#1890ff",
  "slider-dot-border-color": "hsv(0, 0, 91%)",
  "slider-dot-border-color-active": "tint(@primary-color, 50%)",
  "slider-disabled-color": "fade(#000, 25%)",
  "slider-disabled-background-color": "#fff",
  "tree-title-height": "24px",
  "tree-child-padding": "18px",
  "tree-directory-selected-color": "#fff",
  "tree-directory-selected-bg": "#1890ff",
  "collapse-header-padding": "12px 0 12px 40px",
  "collapse-header-bg": "hsv(0, 0, 98%)",
  "collapse-content-padding": "16px",
  "collapse-content-bg": "#fff",
  "message-notice-content-padding": "10px 16px"
}
```

full: https://gist.github.com/gullitmiranda/9ec9944ee31ce77b3366698fa311dd77

**Comment by jean343** [comment]
Some values like `@slider-handle-color-tooltip-open` would work.
However, I am afraid that the more complicated variables `@blue-1` will not work.
Pre-parsing with `less` might resolve those variables.

**Comment by gullitmiranda** [comment]
ye, I'm try without `(reference)` and thong work:

> OBS: i'm updated js-css style to update cache.
```js
const HeaderStyled = styled(Layout.Header)`
  @import "../themes/tmp.less";
  &.light {
    color: @text-color;
    background: @layout-sider-background-light;
  }
```

```less
@import "test.less";
@text-color: blue;
@layout-sider-background-light: cyan;
```

```less
/* @import "test2.less"; */

@test: red;
.test {
  color: @test;
}

@text-color: cyan;
```

**Comment by gullitmiranda** [comment]
never mind, it worked. the cache was not updated as expected.

**Comment by jean343** [comment]
I am glad it works.
I am still looking at a solution to invalidate the cache automatically. Bebel itself does not handle the cache invalidation, it is the responsibility of the bundler.

**Comment by gullitmiranda** [comment]
now I am having problems with the support of `javascriptEnabled` less option 😭 (pr in WIP: https://github.com/jean343/styless/pull/31):

```
Error converting the less syntax for the file: /playground/styless/test/Component/Javascript.test.js `
    @import "import.less";
    color: @echoColor (blue);
    background: @primary-1;
  ` { [Error: JavaScript evaluation error: missing ) after argument list from `echoColor("[props["blue-6"],(props.theme || {})["blue-6"],"#1890ff"].filter(v => v !== void 0)[0]", 1)`]
  message:
   'JavaScript evaluation error: missing ) after argument list from `echoColor("[props["blue-6"],(props.theme || {})["blue-6"],"#1890ff"].filter(v => v !== void 0)[0]", 1)`',
  stack: undefined,
  type: 'Syntax',
  filename:
   '/playground/styless/test/Component/javascript.less',
  index: 324,
  line: 14,
  column: 18,
  callLine: NaN,
  callExtract: undefined,
  extract:
   [ '@primary-color : @blue-6;',
     '@primary-1: color(~`echoColor("@{primary-color}", 1)`);',
     '' ] }
```

**Comment by jean343** [comment]
I do want to support as much of the Less language as possible.
I am afraid that `javascriptEnabled` is deprecated, and I could not find documentation around the ~`` syntax.

I am in favor of passing less arguments to the compiler, and I will approve your PR when complete.

As per your issue, it is transpiling `blue-6` to the props syntax, is `@echoColor (blue)` the correct syntax?

**Comment by gullitmiranda** [comment]
the best syntaxe is `@echoColor(blue)`, and it's work well. the problem is with variables. but already solved this in the PR #31
